include/exadg/grid/grid_utilities.h

```diff
@@ -263,8 +263,16 @@ create_coarse_triangulations(
     unsigned int              level        = fine_triangulation.n_global_levels() - 1;
     std::vector<unsigned int> refine_local = vector_local_refinements;
 
+    // make the last entry of the coarse_triangulations point to the fine_triangulation
+    coarse_triangulations[level].reset(&fine_triangulation, [](auto *) {
+      // empty deleter, since fine_triangulation is an external field
+      // and its destructor is called somewhere else
+    });
+
+    level--;
+
     // undo global refinements
-    for(int refine_global = data.n_refine_global; refine_global >= 0; --refine_global)
+    for(int refine_global = data.n_refine_global - 1; refine_global >= 0; --refine_global)
```

Comment:
`n_refine_global` might be zero, and the data type of  `n_refine_global` should be `unsigned int`, isn't it? Then, I am actually not sure whether this works correctly.

Reply:
Did you test for `n_refine_global = 0`? Could you provide an explanation of how the int-casts will be done here?

Reply:
I double checked just now and did a quick research on `int=unsigned int-unsigned int` operations.

The code works fine for `n_refine_global=0`: The `coarse_triangulations` vector has only one element, which is the fine triangulation.

The reason behind this is for  `data.n_refine_global=0` the loop variable `refine_global`, which is an `int`, becomes `-1` and we don't enter the loop at all. So, regarding functionality, there are no problems for the case `data.n_refine_global=0`. 

If I change the loop variable to `unsigned int refine_global`, then yes, the behavior becomes undefined and `refine_global` becomes a random positive number for negative results.

Reply:
ok. I still wonder whether it would make sense to write the cast to int explicitly (for clarity).

Reply:
Yes, or I could change `refine_global` to `unsigned int` and make changes accordingly. 

I didn't notice that it was `int` until you mentioned it. I agree that it is better to keep the loop variable as `unsigned int`.

Reply:
I agree, keeping the variable `unsigned` is clearer, so I suggest you use the solution proposed last @necioglu . The case `unsigned int(0) - 1` is well-defined due to the implementation of signed integer numbers with two's complement https://en.wikipedia.org/wiki/Two%27s_complement, where `-1=4,294,967,295u`.

Reply:
I would have preferred the previous version (I mean the logic of the for-loop, ignoring data types) with an explicit cast to `int`. From the  current code version it is more difficult to see that we start the loop with the second finest level, not the finest level. The implementation is not longer "intuitive" in my opinion.

Reply:
In my opinion, we should use the cast to `int`, if it is needed to ensure functionality. In this case, since the operation is well-defined, we don't need it. But of course, I can understand that it might reduce confusion. 

I am fine with all the proposed solutions. But I don't know how to weigh in the different arguments. Maybe adding a comment would make the loop clearer?

Reply:
@nfehn @kronbichler How should I proceed here?

Reply:
Yes, this was my intention. It's just about being more explicit. As @necioglu has written, it worked correctly without such a cast. 

Overall, I agree that we need to come to an end in final time. I can live with different potential solutions.